<a href="https://colab.research.google.com/github/10100111/Display-of-HW1/blob/main/44_Ultra_Lite_%D0%92%D0%B2%D0%B5%D0%B4%D0%B5%D0%BD%D0%B8%D0%B5_%D0%B2_%D0%B3%D0%B5%D0%BD%D0%B5%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B5_%D0%B0%D0%BB%D0%B3%D0%BE%D1%80%D0%B8%D1%82%D0%BC%D1%8B_%D0%9F%D1%80%D0%B0%D0%BA%D1%82%D0%B8%D0%BA%D0%B0_1_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

*Теоретический материал:* https://youtu.be/TGO-YPrfs1A

Задание **Ultra Lite**

Макс 10 баллов

Уровень может представлять собой либо тест, либо набор несложных задач, решение которых максимально близко описано в занятии. Это подготовительный уровень, он поможет вам в изучении основ и теоретической части. Обязательно выполните его, если чувствуете неуверенность перед выполнением других уровней ДЗ. Выполнение данного уровня ДЗ необязательно для получения документов о прохождении курса. Дежурные кураторы ответят вам на все вопросы, которые касаются данного уровня, кроме ответов на тесты.

1. С помощью генетического алгоритма найдите точку, где функция от трёх переменных: f = x^2/y + z*(12-z) + 5*y/z  равна 0 (символ ^2 - возведение в квадрат). Интервал изменения каждой переменной x, y, z. Выведите на экран значение этой точки и значение функции в ней.

2. Проведите анализ скорости сходимости до заданной вами точности:

А. Выберите три разные размера популяции.

Б. Выберите три разные коэффициента мутации.

В. Выберите три разных количества выживших.

# Ведение в генетические алгоритмы


Чтобы запускать и редактировать код, сохраните копию этого ноутбука себе (File->Save a copy in Drive...). Свою копию вы сможете изменять и запускать.

In [ ]:
import random as random # Импортируем библиотку генерации случайных значений
import numpy as np # Импортируем библиотеку numpy
import matplotlib.pyplot as plt # Импортируем модуль pyplot библиотеки matplotlib
import matplotlib.colors as colors # Импортируем модуль colors библиотеки matplotlib

# Статический вывод графики (графики отображаются в той ячейке, в которой используется plt.show())
%matplotlib inline 

# Поиск минимума функции

*Разбор данного раздела:* https://youtu.be/dMySAaTqzSI?t=5

In [ ]:
'''
    Функция получения выжившей популяции
        Входные параметры:
        - popul - наша популяция
        - val - текущие значения
        - nsurv - количество выживших
        - reverse - указываем требуемую операцию поиска результата: максимизация или минимизация
'''
def getSurvPopul(
        popul,
        val,
        nsurv,
        reverse
        ):
    newpopul = [] # Двумерный массив для новой популяции    
    sval = sorted(val, reverse=reverse) # Сортируем зачения в val в зависимости от параметра reverse    
    for i in range(nsurv): # Проходимся по циклу nsurv-раз (в итоге в newpopul запишется nsurv-лучших показателей)
        index = val.index(sval[i]) # Получаем индекс i-того элемента sval в исходном массиве val
        newpopul.append(popul[index]) # В новую папуляцию добавляем элемент из текущей популяции с найденным индексом
    return newpopul, sval # Возвращаем новую популяцию (из nsurv элементов) и сортированный список

'''
    Функция получения родителей
        Входные параметры:
        - curr_popul - текущая популяция
        - nsurv - количество выживших
'''
def getParents(
        curr_popul,
        nsurv
        ):   
    indexp1 = random.randint(0, nsurv - 1) # Случайный индекс первого родителя в диапазоне от 0 до nsurv - 1
    indexp2 = random.randint(0, nsurv - 1) # Случайный индекс второго родителя в диапазоне от 0 до nsurv - 1
                                           # random.randint(A, B) генерит случайное целое число включающее ОБЕ границы, поэтому nsurv - 1   
                                           
    botp1 = curr_popul[indexp1] # Получаем первого бота-родителя по indexp1
    botp2 = curr_popul[indexp2] # Получаем второго бота-родителя по indexp2    
    return botp1, botp2 # Возвращаем обоих полученных ботов

'''
    Функция смешивания (кроссинговера) двух родителей
        Входные параметры:
        - botp1 - первый бот-родитель
        - botp2 - второй бот-родитель
        - j - номер компонента бота
'''
def crossPointFrom2Parents(
        botp1,
        botp2, 
        j
        ):
    pindex = random.random() # Получаем случайное число в диапазоне от 0 до 1
    
    # Если pindex меньше 0.5, то берем значения от первого бота, иначе от второго
    if pindex < 0.5:
        x = botp1[j]
    else:
        x = botp2[j]
    return x # Возвращаем значние бота

In [ ]:
n = 100 # Размер популяции
nsurv = 20 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.25 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [0.34172735, 0.99834743, 5.98549904, 9.12409771, 11.88593777]
1   [0.34172735, 0.39922156, 0.99834743, 2.12681514, 3.09205515]
2   [0.34172735, 0.39922156, 0.99834743, 2.12681514, 2.94406671]
3   [0.34172735, 0.37542045, 0.39922156, 0.69958734, 0.99834743]
4   [0.34172735, 0.37542045, 0.39922156, 0.69958734, 0.99834743]
5   [0.34172735, 0.37542045, 0.39922156, 0.69958734, 0.99834743]
6   [0.34172735, 0.37542045, 0.39922156, 0.69958734, 0.99834743]
7   [0.34172735, 0.37542045, 0.39922156, 0.69958734, 0.99834743]
8   [0.34172735, 0.37542045, 0.39922156, 0.69958734, 0.99834743]
9   [0.34172735, 0.37542045, 0.39922156, 0.61138376, 0.69958734]
10   [0.34172735, 0.37542045, 0.39922156, 0.61138376, 0.69958734]
11   [0.34172735, 0.37542045, 0.39922156, 0.61138376, 0.69958734]
12   [0.34172735, 0.37542045, 0.39922156, 0.41048494, 0.61138376]
13   [0.34172735, 0.37542045, 0.39922156, 0.41048494, 0.61138376]
14   [0.18302821, 0.34172735, 0.37542045, 0.39922156, 0.41048494]
15   [0.18302821, 0

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[0.3570535799438221, 0.8328627247465019, 0.7224638712433453, 0.24962868392214582, 0.7149898240342727, 0.8912780883516177, 0.48593042982224993, 0.6092967564034334, 0.6306398850179447, 0.5688776569893006, -0.2948347798299432, 0.5237678583054863]
x = 17.98672301416343 y = 25.263103474100696 z= 13.620876829302174
f = 0.002021251772978516


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = 17.98672301416343 y = 25.263103474100696 z= 13.620876829302174 f = 0.002021251772978516
x = 14.295963466661064 y = 36.61843947048874 z= 13.430977997543698 f = -0.006158886235105854
x = 9.88694052560363 y = 21.29112276716772 z= 12.985437531649003 f = -0.007070592184314606
x = 16.777810629696752 y = 26.656568587452643 z= 13.512157532243684 f = -0.008531921528353692
x = 20.65399110086035 y = 23.804993240740238 z= 13.903889842862048 f = 0.009155469667160077
x = 3.2638872728123625 y = 18.406104683125125 z= 12.622675544604505 f = 0.009830605305421791
x = 11.926794947207252 y = 30.504380809285802 z= 13.224043125838287 f = 0.0100960163164725
x = 8.750825164021872 y = 25.20306192936837 z= 12.982549846203664 f = -0.011089432180822811
x = 11.348290947293538 y = 33.36180488958453 z= 13.241301808947295 f = 0.02139241282040949
x = 11.318982701023742 y = 26.853347660830707 z= 13.139169618489944 f = 0.022148011301137416
x = 22.273686002351347 y = 26.085286627751398 z= 14.018695331631816 f = 0.0233

Ответ:
Лучшие значения точек:

x = 17.98672301416343 

y = 25.263103474100696 

z= 13.620876829302174 

Лучшее знаение функции: 

f = 0.002021251772978516


# Анализ скорости сходимости

## Три разные размера популяции

### 1000 ботов в популяции

x = 5.034902021675882 

y = 5.18222089571994 

z= 12.554045536406026 

f = 0.00022271104923099472

In [ ]:
n = 1000 # Размер популяции
nsurv = 20 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.25 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [0.00635589, 0.05684945, 0.08690542, 0.11897708, 0.50260867]
1   [0.00635589, 0.05684945, 0.08690542, 0.11897708, 0.50260867]
2   [0.00635589, 0.05684945, 0.08690542, 0.1020196, 0.11897708]
3   [0.00635589, 0.05684945, 0.08690542, 0.1020196, 0.11897708]
4   [0.00635589, 0.05684945, 0.08690542, 0.1020196, 0.11897708]
5   [0.00635589, 0.05684945, 0.08690542, 0.1020196, 0.11405632]
6   [0.00635589, 0.05684945, 0.08690542, 0.1020196, 0.11405632]
7   [0.00635589, 0.05684945, 0.08690542, 0.08899936, 0.1020196]
8   [0.00635589, 0.05684945, 0.08690542, 0.08899936, 0.1020196]
9   [0.00635589, 0.05684945, 0.08690542, 0.08899936, 0.1020196]
10   [0.00635589, 0.05684945, 0.08690542, 0.08899936, 0.1020196]
11   [0.00635589, 0.05684945, 0.08481931, 0.08690542, 0.08899936]
12   [0.00635589, 0.0438602, 0.05684945, 0.0792809, 0.08481931]
13   [0.00635589, 0.0438602, 0.05684945, 0.0792809, 0.08481931]
14   [0.00635589, 0.0438602, 0.05684945, 0.0792809, 0.08481931]
15   [0.00635589, 0.00850279, 0.043

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[-0.058731016699902494, -0.396272867828916, 0.7341094443993672, -0.010643882658932291, 1.5856529067908822, 0.9081622841766839, 0.23443876730477015, -0.1315452442000825, 1.1669726536356992, 0.32518607286932993, 0.6184122815907556, 0.1430807881006449]
x = 5.034902021675882 y = 5.18222089571994 z= 12.554045536406026
f = 0.00022271104923099472


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = 5.034902021675882 y = 5.18222089571994 z= 12.554045536406026 f = 0.00022271104923099472
x = 18.868500304620248 y = 19.859372963816785 z= 13.817484286176377 f = 0.00032621830140922015
x = 12.01324799748799 y = 38.324454461821006 z= 13.356179873491257 f = -0.0006025689617867869
x = 15.056661463852404 y = 12.588146150796462 z= 13.656293530077605 f = -0.0006644017902130983
x = 0.5376702462626047 y = 7.393929156175485 z= 12.249489018016757 f = 0.0010413694516850924
x = 3.23707537945711 y = 1.7740380010536585 z= 12.528084055415329 f = -0.001186985453088818
x = 10.87812849894944 y = 16.234641138310806 z= 13.036814729112663 f = -0.0013396788060822118
x = 19.47478901479506 y = 16.872916097274793 z= 14.030525452939049 f = 0.0014732972202482486
x = 18.91918526812419 y = 13.895417108608822 z= 14.164661152773837 f = 0.0025209544805866457
x = 8.17004737787871 y = 31.67950793535258 z= 13.085815322943732 f = 0.0027732427727595876
x = 9.824164188689874 y = 31.867319689195156 z= 13.151274340257597 f

### 500 ботов в популяции

x = 7.422666975135825 

y = 30.872474854253532 

z= 13.044036789937024 

f = 0.00011896302925151758

In [ ]:
n = 500 # Размер популяции
nsurv = 20 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.25 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [0.0325424, 0.11271858, 0.3208139, 0.43907011, 0.62653648]
1   [0.0325424, 0.11271858, 0.14731019, 0.320717, 0.3208139]
2   [0.0325424, 0.11271858, 0.14731019, 0.20610678, 0.2827658]
3   [0.0325424, 0.11271858, 0.14731019, 0.20610678, 0.2827658]
4   [0.01908849, 0.0325424, 0.04225629, 0.11271858, 0.14731019]
5   [0.00541762, 0.01908849, 0.0325424, 0.04225629, 0.07710088]
6   [0.00541762, 0.01908849, 0.0325424, 0.04225629, 0.05650842]
7   [0.00541762, 0.01908849, 0.0325424, 0.04225629, 0.05650842]
8   [0.00541762, 0.01908849, 0.0325424, 0.04225629, 0.05650842]
9   [0.00541762, 0.01908849, 0.0325424, 0.04225629, 0.05650842]
10   [0.00541762, 0.01908849, 0.0325424, 0.04225629, 0.04899832]
11   [0.00541762, 0.01908849, 0.0325424, 0.04225629, 0.04899832]
12   [0.00541762, 0.01908849, 0.0325424, 0.04225629, 0.04899832]
13   [0.00541762, 0.00820076, 0.01908849, 0.0325424, 0.04225629]
14   [0.00541762, 0.00820076, 0.01908849, 0.0255014, 0.0325424]
15   [0.00541762, 0.00820076, 0.01908849, 

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[0.8123484023498133, 0.9097742724280056, 0.7010230994253738, -0.1979513513443102, -0.1736390680614457, 0.7625207648250542, 0.6897949074224166, 0.8134224409586217, 0.850155339864108, 0.8657564883016708, 0.47169582289665973, 0.1259256311839187]
x = 7.422666975135825 y = 30.872474854253532 z= 13.044036789937024
f = 0.00011896302925151758


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = 7.422666975135825 y = 30.872474854253532 z= 13.044036789937024 f = 0.00011896302925151758
x = -0.431845655415767 y = 27.755710923955494 z= 12.842047725503985 f = -0.0003228644841257733
x = -4.246339817178722 y = 28.487782823851685 z= 12.90437167890912 f = 0.0006394254900765617
x = 2.9474684957332133 y = 19.34481228179434 z= 12.640910772859334 f = -0.0009367303093217316
x = 0.8422328341183984 y = 21.35099487970524 z= 12.667732747375789 f = 0.0018785895005457576
x = 28.676623225791566 y = 35.2367702923127 z= 14.457368104073012 f = -0.0028344529544437336
x = 32.35669129756001 y = 16.099977861764074 z= 16.29394779547028 f = 0.0034962901526203183
x = 9.227446587133988 y = 28.588771106940065 z= 13.065574944404487 f = -0.003560256934067141
x = 22.372332690385598 y = 11.401383475809862 z= 15.146575857015598 f = 0.003880536067593976
x = 32.21173325647105 y = 16.31470544983724 z= 16.228437141698315 f = 0.004457319027276085
x = 26.91675492816296 y = 32.91070353004123 z= 14.335914086033885 f =

### 20 ботов в популяции

x = 10.691015792815108 

y = 4.959738858253305 

z= 13.800629920837023 

f = -0.007773979934174946

In [ ]:
n = 20 # Размер популяции
nsurv = 20 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.25 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
1   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
2   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
3   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
4   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
5   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
6   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
7   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
8   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
9   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
10   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
11   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
12   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
13   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.78759785]
14   [0.00777398, 17.7801803, 25.43496078, 34.2167432, 45.

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[0.9918624831556343, 0.2089554195453276, 0.5420292491939493, 0.12936334879973366, 0.2889785678474661, 0.2943890209490111, 0.01748054653072728, 0.1209603888141404, 0.41063435823230954, 0.3251796426944843, 0.7349169381723148, 0.17659711869636574]
x = 10.691015792815108 y = 4.959738858253305 z= 13.800629920837023
f = -0.007773979934174946


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = 10.691015792815108 y = 4.959738858253305 z= 13.800629920837023 f = -0.007773979934174946
x = 23.472965618367073 y = 17.419749369031535 z= 13.503307405787295 f = 17.78018030351984
x = 15.344905385714355 y = 21.424306007535623 z= 11.552329852541286 f = 25.43496077825826
x = 9.126527320178562 y = 26.150860144857305 z= 10.215244976960534 f = 34.21674319907072
x = 9.872356650104889 y = 7.351275228553619 z= 15.867224760409368 f = -45.78759785461608
x = 23.025918779709183 y = 26.78569778332932 z= 5.634604259462856 f = 79.42929325432053
x = 12.001896845934244 y = 32.11459578774816 z= 19.044988978122028 f = -121.2551317287434
x = 13.33278149830712 y = 9.279121128535792 z= 19.785980788315733 f = -132.5510768886571
x = 12.326523081588109 y = 20.030642056391947 z= 20.14872320394801 f = -151.6301339497465
x = 18.019622501900038 y = 13.854396408902304 z= 20.950722400551268 f = -160.78058148805277
x = 17.231554969268146 y = 16.704935635266697 z= 21.203075708082448 f = -173.4194633280989
x = 24.92

## Три разные коэффициента мутации

### Коэффициент мутации = 0,10

x = 5.316526317640403 

y = 20.736020954022845 

z= 12.745167515254245 

f = 0.0006801478923321724

In [ ]:
n = 100 # Размер популяции
nsurv = 20 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.10 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [0.05312551, 1.31726422, 2.14386961, 3.05092308, 3.82087187]
1   [0.05312551, 0.60750618, 1.28968327, 1.31726422, 1.77029799]
2   [0.05312551, 0.52285986, 0.60750618, 1.28968327, 1.31726422]
3   [0.05006175, 0.05312551, 0.52285986, 0.60750618, 1.28968327]
4   [0.05006175, 0.05312551, 0.34943703, 0.49509137, 0.52285986]
5   [0.05006175, 0.05312551, 0.34943703, 0.49509137, 0.52285986]
6   [0.05006175, 0.05312551, 0.34943703, 0.49509137, 0.52285986]
7   [0.05006175, 0.05312551, 0.34943703, 0.49509137, 0.52285986]
8   [0.05006175, 0.05312551, 0.34943703, 0.49509137, 0.52285986]
9   [0.05006175, 0.05312551, 0.2719758, 0.32529817, 0.34943703]
10   [0.05006175, 0.05312551, 0.2719758, 0.32529817, 0.34943703]
11   [0.05006175, 0.05312551, 0.27096531, 0.2719758, 0.32529817]
12   [0.05006175, 0.05312551, 0.27096531, 0.2719758, 0.32529817]
13   [0.05006175, 0.05312551, 0.27096531, 0.2719758, 0.32529817]
14   [0.05006175, 0.05312551, 0.27096531, 0.2719758, 0.27841656]
15   [0.05006175, 0.053125

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[0.49515545131706834, 0.10367225547164396, 0.3880233457787827, 0.016911045247091513, 0.07072198949531877, 0.48180726386887646, 0.946544466533468, 0.35163271919393857, 1.0068880934353512, 0.5685295732876725, 1.0279009153356873, -0.05533510391905369]
x = 5.316526317640403 y = 20.736020954022845 z= 12.745167515254245
f = 0.0006801478923321724


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = 5.316526317640403 y = 20.736020954022845 z= 12.745167515254245 f = 0.0006801478923321724
x = 15.031074284840638 y = 33.941719990326305 z= 13.435353287629562 f = 0.003520621239786692
x = 11.927497183532628 y = 18.48135721102359 z= 13.122835021386983 f = 0.004667221932272092
x = 12.91429430448554 y = 32.39729684813019 z= 13.302090004560126 f = 0.004932270083783052
x = 10.483597381555882 y = 18.69509357523975 z= 13.005116358293254 f = -0.005204650058522553
x = 7.501841371128982 y = 20.06998462942344 z= 12.828814427069165 f = -0.006408454087768689
x = 11.8270030609918 y = 11.340079306183501 z= 13.253959754945205 f = -0.007100898125073307
x = 10.370878253440697 y = 21.83010354090452 z= 13.022722986928798 f = -0.010177422661168833
x = 9.269262314311561 y = 24.28038738997893 z= 12.992371413178942 f = -0.010537630696985545
x = 10.861650290481629 y = 19.61289297191776 z= 13.038949386157181 f = -0.010722831455005633
x = 21.291697247976593 y = 12.160528939550696 z= 14.796350522779242 f = 0.01

### Коэффициент мутации = 0,35

x = 18.642126070906407 

y = 26.010532427480328 

z= 13.67353179247612 

f = -0.010737181767444426

In [ ]:
n = 100 # Размер популяции
nsurv = 20 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.35 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [0.90839869, 3.41838276, 7.33759929, 11.26253686, 11.72052907]
1   [0.90839869, 3.41838276, 5.82198733, 7.33759929, 9.00664024]
2   [0.63697643, 0.90839869, 3.41838276, 5.82198733, 6.72496969]
3   [0.63697643, 0.90839869, 1.38797809, 1.72904569, 2.87751795]
4   [0.63697643, 0.90839869, 1.38797809, 1.72904569, 2.87751795]
5   [0.63697643, 0.90839869, 1.14605073, 1.38797809, 1.56851573]
6   [0.63697643, 0.90839869, 1.14605073, 1.38797809, 1.56851573]
7   [0.63697643, 0.90839869, 1.14605073, 1.38797809, 1.56851573]
8   [0.63697643, 0.90839869, 1.14605073, 1.15747149, 1.38797809]
9   [0.58181512, 0.63697643, 0.90839869, 1.14605073, 1.15747149]
10   [0.24193993, 0.58181512, 0.63697643, 0.90839869, 1.14605073]
11   [0.24193993, 0.45895053, 0.58181512, 0.63697643, 0.90839869]
12   [0.24193993, 0.45895053, 0.58181512, 0.63697643, 0.90839869]
13   [0.24193993, 0.45895053, 0.58181512, 0.61131538, 0.63697643]
14   [0.24193993, 0.45895053, 0.58181512, 0.61131538, 0.63697643]
15   [0.24193993, 

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[1.8772947761197993, 1.1293146356566297, -0.589338360635772, 0.6804656689144472, 0.3866682535887259, 1.0235669342931903, 1.2720183271215963, 0.31143384924838735, 0.18519378380969972, 0.8077093724657808, -0.2499729820265802, 0.47798083866413277]
x = 18.642126070906407 y = 26.010532427480328 z= 13.67353179247612
f = -0.010737181767444426


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = 18.642126070906407 y = 26.010532427480328 z= 13.67353179247612 f = -0.010737181767444426
x = 31.3704643441841 y = 23.593371991022032 z= 15.244658452283439 f = -0.014344996560754275
x = 13.399794321049171 y = 18.369401721410252 z= 13.260569626593956 f = -0.014895945142115785
x = 18.934585386994627 y = 15.109731399188128 z= 14.069435261048959 f = -0.018427702630474663
x = 19.02583774454599 y = 18.997781654509158 z= 13.866629970829006 f = 0.020247952253548362
x = 3.9758936810973253 y = 22.184640578777458 z= 12.737842141734905 f = 0.02219898389530961
x = 27.01328842499295 y = 16.154400402339125 z= 15.299430536129435 f = -0.028542173920425462
x = 11.378017977558393 y = 7.125947878955663 z= 13.539091545916715 f = -0.03897323732476865
x = 21.42012602393989 y = 29.52495433913273 z= 13.882477656646532 f = 0.040574016827671144
x = -11.387902580822686 y = 14.311047208352052 z= 13.10492789343185 f = 0.04201077794026631
x = 12.954755278768506 y = 52.02014681918795 z= 13.638102371394297 f = -0.0

### Коэффициент мутации = 0,80

x = 9.85087942811564 

y = -29.45379773986025 

z= 4.7353611144690335 

f = 0.006199561422430833

In [ ]:
n = 100 # Размер популяции
nsurv = 20 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.80 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [1.39959295, 8.34585388, 12.1601859, 13.26996206, 14.04052893]
1   [1.39959295, 2.13366327, 8.34585388, 9.95969767, 10.15026956]
2   [1.39959295, 2.13366327, 3.59843911, 4.97593738, 8.34585388]
3   [0.16979035, 1.39959295, 1.83529543, 2.13366327, 3.59843911]
4   [0.16979035, 1.39959295, 1.83529543, 2.13366327, 3.59843911]
5   [0.1554997, 0.16979035, 1.39959295, 1.83529543, 2.13366327]
6   [0.1554997, 0.16979035, 0.59096209, 1.39959295, 1.83529543]
7   [0.1554997, 0.16979035, 0.59096209, 1.39959295, 1.83529543]
8   [0.1554997, 0.16979035, 0.20729033, 0.59096209, 1.39959295]
9   [0.1554997, 0.16979035, 0.20729033, 0.40765394, 0.59096209]
10   [0.1554997, 0.16979035, 0.20729033, 0.40765394, 0.59096209]
11   [0.1554997, 0.16979035, 0.20729033, 0.40765394, 0.59096209]
12   [0.1554997, 0.16979035, 0.20729033, 0.40765394, 0.59096209]
13   [0.1554997, 0.16979035, 0.20729033, 0.40765394, 0.59096209]
14   [0.07354978, 0.1554997, 0.16979035, 0.20729033, 0.40765394]
15   [0.07354978, 0.1554997

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[1.4793291246019016, 0.1900961320729707, 0.5385277048008247, 0.08143170380562548, 1.4258992545497833, -0.7813082247847203, -0.6841793532703223, -0.8052544935113284, -0.4671704544070991, -1.4798066767115043, 2.0420139840514424, -0.31274299552323076]
x = 9.85087942811564 y = -29.45379773986025 z= 4.7353611144690335
f = 0.006199561422430833


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = 9.85087942811564 y = -29.45379773986025 z= 4.7353611144690335 f = 0.006199561422430833
x = 15.94310601349137 y = -37.791806206191765 z= 8.782935605028129 f = 0.015066905834835609
x = -13.654946455282566 y = -42.73009450496973 z= 8.605482321952554 f = 0.02059830457556444
x = -8.526851993176024 y = -41.29940742398411 z= -3.584724145332494 f = -0.022705412275449532
x = -15.857326890064657 y = -29.298046507806465 z= 5.415443675736189 f = 0.025187891360491932
x = 0.91118527777919 y = -35.40876061717384 z= 10.33897003102085 f = 0.02596085245851043
x = -14.525287703279027 y = -36.65823580599471 z= 9.225407760621508 f = -0.0267715005504634
x = 20.003392985813377 y = -30.448713485082497 z= 6.9331064359476855 f = 0.029085198625150355
x = -1.4246974590481738 y = -38.43799739556475 z= -3.5190820061418933 f = -0.05205737412717326
x = -13.819153895461351 y = -28.05784262138168 z= 4.989038627444432 f = 0.052207330918900396
x = 16.44090833350989 y = -26.848503405217986 z= 5.271940088827417 f = -0.

## Три разных количества выживших

### 10% выживших

x = -2.654529971813864 

y = -1.5491613523604948 

z= 1.0755252110053108 

f = -0.0009440395212791941

In [ ]:
n = 100 # Размер популяции
nsurv = 10 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.25 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [1.21598425, 7.35821169, 16.27756913, 17.98389483, 19.903444]
1   [1.21598425, 2.76365192, 3.02407049, 4.12204634, 4.43729374]
2   [0.66909149, 1.21598425, 1.28848126, 1.81025147, 2.76365192]
3   [0.66909149, 1.21598425, 1.28848126, 1.81025147, 1.88041377]
4   [0.32599856, 0.66909149, 1.21598425, 1.28848126, 1.81025147]
5   [0.32599856, 0.66909149, 1.05052074, 1.21598425, 1.28848126]
6   [0.32599856, 0.60018468, 0.60432023, 0.66909149, 1.05052074]
7   [0.32599856, 0.54398775, 0.60018468, 0.60432023, 0.66909149]
8   [0.32599856, 0.54398775, 0.60018468, 0.60432023, 0.66909149]
9   [0.08520884, 0.32599856, 0.54398775, 0.60018468, 0.60432023]
10   [0.08520884, 0.32599856, 0.54398775, 0.60018468, 0.60432023]
11   [0.08520884, 0.32599856, 0.34630866, 0.54398775, 0.60018468]
12   [0.08520884, 0.20522426, 0.32599856, 0.34630866, 0.54398775]
13   [0.05543201, 0.08520884, 0.20522426, 0.32599856, 0.34630866]
14   [0.05543201, 0.08520884, 0.20522426, 0.32599856, 0.34630866]
15   [0.05543201, 0

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[0.09702417950614917, -0.03814871991311547, 0.06067314741572627, -0.12670168103646795, 0.5650862826759875, 0.7848369079564013, -0.0840672941381499, -0.20791036933747958, 0.21353244510591474, 0.6437798425727956, -0.29312891508479627, 0.022975308155335217]
x = -2.654529971813864 y = -1.5491613523604948 z= 1.0755252110053108
f = -0.0009440395212791941


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = -2.654529971813864 y = -1.5491613523604948 z= 1.0755252110053108 f = -0.0009440395212791941
x = 2.9297072394072208 y = -2.4099859155783783 z= 1.2362866363114238 f = -0.0013467737038759253
x = 16.328095439801757 y = -14.556155469423645 z= 9.114570355135246 f = -0.0013868631591043012
x = 16.850730552519693 y = -13.624282609406162 z= 8.708927435622112 f = -0.0015617757988106717
x = 1.2154882363309472 y = -21.12219235566218 z= 3.5364049201414978 f = -0.0031802771405011754
x = 15.399764194063456 y = -8.993130188345031 z= 8.000640379828559 f = 0.006748574595479617
x = 16.202710762920233 y = -14.311984439342705 z= 9.135332755999082 f = -0.006841625395439976
x = 12.60243592457746 y = -18.446801858096315 z= 10.291769675053477 f = 0.009095854725506669
x = 14.93419992164155 y = -9.936308757353956 z= 4.450453902275651 f = -0.01034036896951207
x = 15.557242096399683 y = -11.802320402031933 z= 4.4800177705186535 f = 0.01067727006557817
x = 21.886192039615292 y = -2.356933433959636 z= 5.000923877

### 50% выживших

x = 18.04473851455305 

y = 4.015872444889357 

z= 16.875086308179302 

f = 0.003786809461588714

In [ ]:
n = 100 # Размер популяции
nsurv = 50 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.25 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [0.85656843, 3.45893659, 3.70211052, 10.80993952, 16.56220181]
1   [0.85656843, 0.92607729, 3.45893659, 3.70211052, 5.30607211]
2   [0.85656843, 0.92607729, 2.78041843, 3.45893659, 3.70211052]
3   [0.00378681, 0.85656843, 0.92607729, 2.78041843, 3.01945042]
4   [0.00378681, 0.85656843, 0.92607729, 1.09359981, 1.66019533]
5   [0.00378681, 0.85656843, 0.92607729, 1.09359981, 1.66019533]
6   [0.00378681, 0.85656843, 0.92607729, 1.09359981, 1.66019533]
7   [0.00378681, 0.85656843, 0.92607729, 1.09359981, 1.66019533]
8   [0.00378681, 0.85656843, 0.92607729, 1.09359981, 1.63875874]
9   [0.00378681, 0.43063215, 0.85656843, 0.92607729, 1.09359981]
10   [0.00378681, 0.43063215, 0.85656843, 0.92607729, 1.09359981]
11   [0.00378681, 0.43063215, 0.85656843, 0.92607729, 0.95130627]
12   [0.00378681, 0.43063215, 0.85656843, 0.92607729, 0.95130627]
13   [0.00378681, 0.43063215, 0.85656843, 0.92607729, 0.95130627]
14   [0.00378681, 0.43063215, 0.85656843, 0.92607729, 0.95130627]
15   [0.00378681, 

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[0.5906598193264099, 0.2059457988863696, 1.2098573178213963, 0.17303106090323317, 0.2745003559724281, 0.1745764355518295, 0.3670148501776452, -0.03206634362474681, 0.9266856282973348, 0.7189031960679615, -0.034934169522039094, 0.508129055790502]
x = 18.04473851455305 y = 4.015872444889357 z= 16.875086308179302
f = 0.003786809461588714


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = 18.04473851455305 y = 4.015872444889357 z= 16.875086308179302 f = 0.003786809461588714
x = 11.15972380636141 y = 23.043589170238818 z= 13.085017249901178 f = 0.012379984841892622
x = 15.462331189319592 y = 5.902293977609381 z= 14.860331610846508 f = -0.012644423553177075
x = 17.84206668233391 y = 33.45305780331432 z= 13.602076913062794 f = 0.021468343624356123
x = 18.67425066160437 y = 13.350343303893997 z= 14.176297165041811 f = -0.021901715000828048
x = 4.510245242823028 y = 37.55125360070988 z= 13.132515592767 f = -0.03400469078843038
x = 19.105890461213875 y = 11.785300523827143 z= 14.431678630964367 f = -0.036310475056668956
x = 18.026392465151467 y = 32.160182330791926 z= 13.612712278418563 f = -0.0366982513255536
x = 14.73226729112475 y = 11.75523869076749 z= 13.663134076673899 f = 0.04141717777084253
x = 16.23100067643671 y = 32.49530131038971 z= 13.49021334904119 f = 0.04791581023238933
x = 23.29001805130027 y = 28.67855658625413 z= 14.072378627391112 f = -0.05967919829966

### 70% выживших

x = 0.19372288854556174 

y = 12.109780604104756 

z= 12.394058449933095 

f = 0.00443245264327885

In [ ]:
n = 100 # Размер популяции
nsurv = 70 # Количество выживших (столько лучших переходит в новую популяцию)
nnew = n - nsurv # Количество новых (столько новых ботов создается)
# l = 8 # Длина бота
l = 12 # Длина бота
epohs = 500 # Количество эпох

mut = 0.25 # Коэфициент мутаций

# popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 8 компонентов каждый
popul = [] # Двумерный массив популяции, размерностью [n, l]. 100 ботов по 12 компонентов каждый
val = [] # Одномерный массив значений этих ботов

for i in range(n): # Проходим по всей длине популяции
    popul.append([]) # Создаем пустого бота
    for j in range(l): # Проходим по всей длине бота
        popul[i].append(random.random()) # В каждый компонент бота записываем рандомное значение 

for it in range(epohs): # Пробегаемся по всем эпохам
    val = [] # Создаем пустой список для значений ботов

    for i in range(n): # Проходим по всей длине популяции
        bot = popul[i] # Берем очередного бота
    
        # в каждом боте должны быть закодированы парные значения x-y
        x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # первые 4 значения отводим для Х
        y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # вторые 4 значения - для Y
        z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # третьи 4 значения - для Z
        # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # для кодирования применяем эту функцию
        f = x**2/y + z*(12-z) + 5*y/z # для кодирования применяем эту функцию
        val.append(abs(f)) # добавляем модуль значения в список на эпоху
                           # в этой задаче будем искать 0 функции
    
    newpopul, sval = getSurvPopul(popul, val, nsurv, 0) # Получаем новую популяцию и сортированный список значнией
    print(it, " ", [round(s,8) for s in sval[0:5]]) # Выводим 5 лучших ботов
 
    for i in range(nnew): # Проходимся в цикле nnew-раз 
        botp1, botp2 = getParents(newpopul, nsurv) # Из newpopul(новой популяции) получаем двух случайных родителей-ботов
        newbot = [] # Массив для нового бота
        # проходимся по длине бота и осуществляем смешивание/скрещивание от родителей
        for j in range(l): # Проходим по всей длине бота
            x = crossPointFrom2Parents(botp1, botp2, j) # Получаем значение для j-ого компонента бота
            x += mut*(2*random.random() - 1.0) #Добавялем к значению бота случайную величину, зависящую от коэфециента мутации        
            newbot.append(x) # Добавялем новое значение в бота      
        newpopul.append(newbot) # Добавляем бота в новую популяцию 
        #(таким образом к nsurv-лучших ботов предыдующей популяции добавится nnew-новых ботов)
    
    popul = newpopul # Записываем в popul посчитанную новую популяцию

0   [0.78057201, 8.88360789, 10.59576511, 14.55274112, 17.59014765]
1   [0.64697122, 0.78057201, 8.88360789, 8.9877031, 10.59576511]
2   [0.48655683, 0.64697122, 0.78057201, 8.34695937, 8.88360789]
3   [0.48655683, 0.64697122, 0.78057201, 5.71711456, 6.88122411]
4   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.96178231]
5   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.96178231]
6   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.96178231]
7   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.96178231]
8   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.96178231]
9   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.81363753]
10   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.81363753]
11   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.81363753]
12   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.81363753]
13   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.81363753]
14   [0.21088547, 0.48655683, 0.64697122, 0.78057201, 0.81363753]
15   [0.21088547,

In [ ]:
bot = popul[0] # Берем первого (лучшего) бота в финальной популяции
x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
# f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции
# print(bot) # Выводим бота (8 компонент)
print(bot) # Выводим бота (12 компонент)
# print("x =",x, "y =", y) # Выводим значения x и y
print("x =",x, "y =", y, "z=", z) # Выводим значения x,y,z
print("f =",f) # Выводим значение функции (которое должно быть максимально близко к 0 по условию задачи)


[0.8211060056128758, 1.3238615773049693, 0.1287897867441008, -0.34138355484132676, 0.006305918466946392, 0.5513357621085044, 0.1315672887680711, 0.32124491949658307, 0.0455356966676852, 0.22251188025554813, 0.9281777439751606, 0.07816743648944252]
x = 0.19372288854556174 y = 12.109780604104756 z= 12.394058449933095
f = 0.00443245264327885


In [ ]:
# Выведем значения 20 лучших ботов
for i in range(20):
    bot = popul[i] # Берем очередного бота
    x = bot[0] + 5*bot[1] + 10*bot[2] + 25*bot[3] # Считаем x
    y = bot[4] + 5*bot[5] + 10*bot[6] + 25*bot[7] # Считаем y
    z = bot[8] + 5*bot[9] + 10*bot[10] + 25*bot[11] # Считаем z        
    # f = x*(y-5)/(17+x) + y + 0.25*y*y/x + 19/y # Считаем значение функции
    f = x**2/y + z*(12-z) + 5*y/z # Считаем значение функции    
    # print("x =",x, "y =", y, "f =", f) # Выводим посчитанные значения
    print("x =",x, "y =", y, "z=", z, "f =", f) # Выводим посчитанные значения

x = 0.19372288854556174 y = 12.109780604104756 z= 12.394058449933095 f = 0.00443245264327885
x = 2.2884692558929327 y = 12.600883255647755 z= 12.439299774585226 f = 0.015980248921763085
x = 6.84037264853206 y = 38.90639919029216 z= 13.207641400019398 f = -0.018700646268179355
x = 5.422338097382603 y = 7.26592369157683 z= 12.550566521900183 f = 0.03126406949776195
x = -2.2037527320678123 y = 27.685725291712053 z= 12.849511525324345 f = 0.03267339271807046
x = 6.047667687289966 y = 20.781340470185405 z= 12.771903574809281 f = 0.036848844686884874
x = -0.008521276662689647 y = 21.00201053434814 z= 12.652920052368918 f = 0.03793241255895374
x = 14.147862317350102 y = 21.13837243602027 z= 13.30563089633619 f = 0.04028126902373419
x = 0.2848646153345973 y = 14.251171230808586 z= 12.456124897388896 f = 0.044693088780875634
x = 7.143062287984375 y = 34.076440322135184 z= 13.102893015129279 f = 0.049634159815019174
x = 17.996204272240366 y = 34.16868019163149 z= 13.620761198178712 f = -0.054764

# Выводы

In [ ]:
import pandas as pd

In [ ]:
# берем модуль значения функции для поиска значения, ближайшего к 0
df = pd.DataFrame({'Наименование исследования': ['Задание №1', '1000 ботов', '500 ботов', '20 ботов', 'Коэф.мутации = 0,10', 'Коэф.мутации = 0,35', 'Коэф.мутации = 0,80', '10 выживших', '50 выживших', '70 выживших'], 'Результат минимума функции': [0.0020, 0.0002, 0.0001, 0.0077, 0.0006, 0.0107, 0.0061, 0.0009, 0.0037, 0.004]})
df

Наименование исследования  Результат минимума функции
0                Задание №1                      0.0020
1                1000 ботов                      0.0002
2                 500 ботов                      0.0001
3                  20 ботов                      0.0077
4       Коэф.мутации = 0,10                      0.0006
5       Коэф.мутации = 0,35                      0.0107
6       Коэф.мутации = 0,80                      0.0061
7               10 выживших                      0.0009
8               50 выживших                      0.0037
9               70 выживших                      0.0040

In [ ]:
df.sort_values(by=['Результат минимума функции'])

Наименование исследования  Результат минимума функции
2                 500 ботов                      0.0001
1                1000 ботов                      0.0002
4       Коэф.мутации = 0,10                      0.0006
7               10 выживших                      0.0009
0                Задание №1                      0.0020
8               50 выживших                      0.0037
9               70 выживших                      0.0040
6       Коэф.мутации = 0,80                      0.0061
3                  20 ботов                      0.0077
5       Коэф.мутации = 0,35                      0.0107

Самый лучший результат дали исследования с количеством ботов в популяции по 1000 и по 500. 

Также хороший результат получен с минимальным коэффициентом мутации. 

Чуть менее хороший результат получен от исследования с 10 выжившими ботами, приток новых ботов в этих популяциях на 90% дал более эффективную среду для естественного отбора лучших ботов.

На порядок хуже результат Задания №1. 

Далее видно, что с увеличением выживших популяция вырождается и результат ухудшается. 

Маленькая популяция ботов и повышение коэффициента мутации дали наихудшие результаты.

